src/Enzo/enzo_EnzoRiemannHLL.hpp

```diff
@@ -0,0 +1,299 @@
+
+    return std::sqrt(0.5 * (tilde_a2 + tilde_va2 +
+                            std::sqrt(std::pow(tilde_a2 + tilde_va2, 2) -
+                                      4 * tilde_a2 * tilde_vai2)));
+  }
+};
+
+template<class inputLUT>
+struct DavisWavespeed
+{
+  /// @class    DavisHLLEWavespeed
+  /// @ingroup  Enzo
+  /// @brief    [\ref Enzo] Encapsulates one of the wavespeed estimators
+  ///           described by Davis 1998. It was used in Enzo's Runge-Kutta MHD
+  ///           HLL Riemann solver in Riemann_HLL_MHD.C which was (written by
+  ///           Peng Wang)
+  ///
+  /// Toro, describes this version as the "most well-known approach." He goes
+  /// on to describe it as "exceedling simple" and not recommended for
+  /// practical calculations. He attributes this to the scheme from Davis 1988
+public:
+
+  using LUT = EnzoRiemannLUT<inputLUT>;
+
+  void operator()(const lutarray<LUT> wl, const lutarray<LUT> wr,
+                  const lutarray<LUT> Ul, const lutarray<LUT> Ur,
+                  enzo_float pressure_l, enzo_float pressure_r,
+                  const enzo_float gamma,
+                  enzo_float *bp, enzo_float *bm) const noexcept
+  {
+    // Should be relatively easy to adapt and make compatible with barotropic
+    // eos
+    ERROR("EnzoHLLEWavespeed","This hasn't been tested yet");
+
+    // compute left and right fast magnetosonic speed assuming that cos2 = 1
+    // not sure why cos2 = 1 was chosen, (cos2 = 0, would yield faster speeds)
+    using enzo_riemann_utils::fast_magnetosonic_speed;
+    enzo_float c_l = fast_magnetosonic_speed<LUT>(wl, pressure_l, gamma, 1);
+    enzo_float c_r = fast_magnetosonic_speed<LUT>(wr, pressure_r, gamma, 1);
+
+    enzo_float lp_l = wl[LUT::velocity_i] + c_l;
+    enzo_float lm_l = wl[LUT::velocity_i] - c_l;
+    enzo_float lp_r = wr[LUT::velocity_i] + c_r;
+    enzo_float lm_r = wr[LUT::velocity_i] - c_r;
+
+    // The following is equivalent to the enzo version except bm has been
+    // multiplied by -1
+    *bp = std::fmax(lp_l, lp_r);
+    *bm = std::fmin(lm_l, lm_r);
+  }
+};
+
+
+template <class WaveSpeedFunctor>
+struct HLLImpl
+{
+  /// @class    EnzoRiemannHLLE
+  /// @ingroup  Enzo
+  /// @brief    [\ref Enzo] Encapsulates operations of HLLE approximate Riemann
+  /// Solver. 
+public:
+
+  using LUT = typename WaveSpeedFunctor::LUT;
+
+  lutarray<LUT> operator()
+  (const lutarray<LUT> flux_l, const lutarray<LUT> flux_r,
+   const lutarray<LUT> prim_l, const lutarray<LUT> prim_r,
+   const lutarray<LUT> cons_l, const lutarray<LUT> cons_r,
+   enzo_float pressure_l, enzo_float pressure_r,
+   bool barotropic_eos, enzo_float gamma, enzo_float isothermal_cs,
+   enzo_float &vi_bar) const noexcept
+  { 
+    // there is no scratch_space
+    WaveSpeedFunctor wave_speeds;
+    enzo_float bp, bm;
+
+    // Compute wave speeds
+    wave_speeds(prim_l, prim_r, cons_l, cons_r, pressure_l, pressure_r, gamma,
```

Comment:
I realize that might not be exactly what you're talking about (because `static` means a million things in c++). But I think what you're suggesting is defining a traditional "c-style" inline function and labeling it with the `static` keyword to inform the compiler that this implementation function is only to be used by the local translation unit. 

I think there might be a bit of issue with that. According to this [stackoverflow post](https://stackoverflow.com/a/1174193/4538758), traditional functions don't play well with templates. Apparently the function may decay down to a function pointer, blocking optimizations.

But there's a work-around for this (which you might have also been referring to). We could instead implement these `wave_speeds` functions by defining them as static methods for different structs (we would would have one struct per type of wavespeed). I think I originally had this implementation. I'm more than happy to go back to it (I had shifted to using functors because I was worried that it was too convoluted for people to follow - but that doesn't really seem to be an issue).

If we really to have a similar result to a having a traditional static function, we could place the structs [anonymous namespaces](https://llvm.org/docs/CodingStandards.html#anonymous-namespaces). But I don't really know how much that gets us.

Let me know what your thoughts are on this

Reply:
I do mean a static inline member method, so that the member function can be directly called from the struct definition without creating an object. They would still essentially be functor structs as you have now, but I think you would have to move away from `operator::()` since you couldn't make it static. From what I see, none of the wavespeed structs have data members, so they don't really need to be created as objects.

I also think this should be left for later, not in this PR. I'm not sure if it would be a huge performance difference either.

Reply:
I originally had exactly what you're were describing (I had only moved away from it because I found it difficult to describe in the documentation - but I'm happy to go back to it in the future).